### PR TITLE
feat!: update helm chart to support distributed mode and 3.0

### PR DESCRIPTION
**Pull request by slim-bean**
**TODO**
- [x] fix toplogy spread constraints in zone aware
- [ ] fix SSD to to handle zone aware migration (honor not writing to "default-zone")
- [ ] distributor also needs to have config for not writing "default-zone"
- [x] replica count for zone aware needs to be a divide by 3
- [x] add rollout operator
- [x] add memcached
- [x] make sure configmap isn't a secret (or make configurable)
- [ ] verify query-scheduler and query-frontend addresses are set correctly in distributed
- [x] add configs to persist tokens and not unregister on shutdown
- [ ] add mutating webhook and labels so rollout operator can control scaledown

### CODE REVIEW

File: production/helm/loki/templates/ingester/statefulset-ingester-zone-a.yaml [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    {{- include "loki.ingesterLabels" . | nindent 4 }}
+    app.kubernetes.io/part-of: memberlist
+    rollout-group: ingester
+    name: ingester-zone-a
+  annotations:
+    rollout-max-unavailable: "{{ include "loki.ingester.maxUnavailable" (dict "ctx" . "replicas" $replicas)}}"
+    {{- with .Values.loki.annotations }}
+    {{- toYaml . | nindent 4 }}
+    {{- end }}
+    {{- with .Values.ingester.zoneAwareReplication.zoneA.annotations }}
```
Comment (yoyosir) [inline production/helm/loki/templates/ingester/statefulset-ingester-zone-a.yaml:19]: Can we also have this in pod annotation in spec.template.metadata.annotations?

File: production/helm/loki/templates/memcached/_memcached-statefulset.tpl [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      {{- end }}
+      securityContext:
+        {{- toYaml $.ctx.Values.memcached.podSecurityContext | nindent 8 }}
+      initContainers:
+        {{- toYaml .initContainers | nindent 8 }}
+      nodeSelector:
+        {{- toYaml .nodeSelector | nindent 8 }}
+      affinity:
+        {{- toYaml .affinity | nindent 8 }}
+      toplogySpreadConstraints:
```
Comment (yoyosir) [inline production/helm/loki/templates/memcached/_memcached-statefulset.tpl:68]: typo

File: production/helm/loki/values.yaml [hunk 3]
```diff
@@ -1138,386 +1987,740 @@ backend:
...
+  initContainers: []
+  # -- Annotations for the results-cache pods
+  annotations: {}
+  # -- Node selector for results-cache pods
+  nodeSelector: {}
+  # -- Affinity for results-cache pods
+  affinity: {}
+  # -- topologySpreadConstraints allows to customize the default topologySpreadConstraints. This can be either a single dict as shown below or a slice of topologySpreadConstraints.
+  # labelSelector is taken from the constraint itself (if it exists) or is generated by the chart using the same selectors as for services.
+  topologySpreadConstraints: {}
```
Comment (yoyosir) [inline production/helm/loki/values.yaml:2556]: I think default needs to be []? Same for other places in this file.

File: production/helm/loki/values.yaml [hunk 4]
```diff
@@ -1138,386 +1987,740 @@ backend:
...
+  initContainers: []
+  # -- Annotations for the chunks-cache pods
+  annotations: {}
+  # -- Node selector for chunks-cache pods
+  nodeSelector: {}
+  # -- Affinity for chunks-cache pods
+  affinity: {}
+  # -- topologySpreadConstraints allows to customize the default topologySpreadConstraints. This can be either a single dict as shown below or a slice of topologySpreadConstraints.
+  # labelSelector is taken from the constraint itself (if it exists) or is generated by the chart using the same selectors as for services.
+  topologySpreadConstraints: {}
```
Comment (yoyosir) [inline production/helm/loki/values.yaml:2639]: same here

Comment (slim-bean) [reply production/helm/loki/values.yaml]: fixed!

File: production/helm/loki/values.yaml (hunk 3 above)
Comment (slim-bean) [reply production/helm/loki/values.yaml]: fixed!

File: production/helm/loki/templates/memcached/_memcached-statefulset.tpl (hunk 2 above)
Comment (slim-bean) [reply production/helm/loki/templates/memcached/_memcached-statefulset.tpl]: fixed!

File: docs/sources/setup/upgrade/upgrade-to-6x/index.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+`generatedConfigObjectName` also can be used to control the name of the config object created by the chart.
+
+This gives greater flexibility in using the chart to still generate a config object but allowing for another process to load and mutate this config into a new object which can be loaded by Loki and `configObjectName`
+
+#### Monitoring
+
+After some consideration of how this chart works with other charts provided by Grafana, we decided to deprecate the monitoring sections of this chart and take a new approach entirely to monitoring Loki, Mimir and Tempo with the [Meta Monitoring Chart](https://github.com/grafana/meta-monitoring-chart)
+
+Reasons:
+  * There were conflicts with this chart and the Mimir chart both installing the Agent Operator
```
Comment (JStickler) [inline docs/sources/setup/upgrade/upgrade-to-6x/index.md:47]:
```suggestion
  * There were conflicts with this chart and the Mimir chart both installing the Agent Operator.
```

File: docs/sources/setup/upgrade/upgrade-to-6x/index.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  * The Agent Operator is deprecated
```
Comment (JStickler) [inline docs/sources/setup/upgrade/upgrade-to-6x/index.md:48]:
```suggestion
  * The Agent Operator is deprecated.
```

File: docs/sources/setup/upgrade/upgrade-to-6x/index.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  * The dependency on the Prometheus operator is not one we are able to support well.
+
+The [Meta Monitoring Chart](https://github.com/grafana/meta-monitoring-chart) improves several things here by allowing for installing a clustered Grafana Agent which can send metrics, logs, and traces to Grafana Cloud, or the ability to install a monitoring only installation of Loki, Mimir, Tempo, and Grafana locally.
+
+The monitoring sections of this chart still exist but are disabled by default.
+
+If you wish to continue using the self monitoring features you should use these configs, but please do note a future version of this chart will remove this capability completely:
```
Comment (JStickler) [inline docs/sources/setup/upgrade/upgrade-to-6x/index.md:55]:
```suggestion
If you wish to continue using the self monitoring features you should use the following configuration, but please do note a future version of this chart will remove this capability completely:
```

File: docs/sources/setup/upgrade/upgrade-to-6x/index.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  enabled: true
+  selfMonitoring:
+    enabled: true
+    grafanaAgent:
+      installOperator: true
+```
+
+#### Memcached is included and enabled by default
+
+Caching is crucial to the proper operation of Loki and Memcached is now included in this chart and enabled by default for the `chunksCache` and `resultsCache`
```
Comment (JStickler) [inline docs/sources/setup/upgrade/upgrade-to-6x/index.md:68]:
```suggestion
Caching is crucial to the proper operation of Loki and Memcached is now included in this chart and enabled by default for the `chunksCache` and `resultsCache`.
```

File: docs/sources/setup/upgrade/upgrade-to-6x/index.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+If you are already running Memcached separately you can remove your existing installation and use the Memcached deployments built into this chart.
+
+##### Single Binary
+
+Memcached also deploys for the Single Binary, but this may not be desired in resource constrained environments.
+
+You can disable it with the following.
```
Comment (JStickler) [inline docs/sources/setup/upgrade/upgrade-to-6x/index.md:76]:
```suggestion
You can disable it with the following configuration:
```

File: docs/sources/setup/upgrade/upgrade-to-6x/index.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  enabled: false
+resultsCache:
+  enabled: false
+```
+
+With these caches disabled, Loki will return to defaults which enables an in-memory results and chunks cache, so you will still get some caching.
+
+#### Distributed mode
+
+This chart introduces the ability to run Loki in distributed, or microservices mode. Separate instructions on how to enable this as well as how to migrate from the existing community chart will be coming shortly!
```
Comment (JStickler) [inline docs/sources/setup/upgrade/upgrade-to-6x/index.md:89]:
```suggestion
This chart introduces the ability to run Loki in distributed, or [microservices mode](https://grafana.com/docs/loki/latest/get-started/deployment-modes/#microservices-mode). Separate instructions on how to enable this as well as how to migrate from the existing community chart will be coming shortly.
```

**Review by trevorwhitney — APPROVED** [review]

**Comment by messiahUA** [comment]
May I ask why the simple scalable check was removed from the helm chart for "loki.querySchedulerAddress" ?

File: production/helm/loki/values.yaml [hunk 11]
```diff
@@ -357,21 +416,32 @@ loki:
...
   analytics: {}
   # --  Optional querier configuration
+  query_range: {}
+  # --  Optional querier configuration
   querier: {}
   # --  Optional ingester configuration
   ingester: {}
   # --  Optional index gateway configuration
   index_gateway:
-    mode: ring
```
Comment (lindeskar) [inline production/helm/loki/values.yaml:365]: Hi,
I'm curious to the reason behind this change. Is ring mode not recommended for Index Gateway? Please help me understand. 🙂

**Pull request merged**